gulpfile.js/changelog.js

```diff
@@ -0,0 +1,380 @@
+ * @returns {Promise<CommitInfo>}
+ *
+ * @typedef {{ message: string, hash: string, changes: CommitChange[] }} CommitInfo
+ * @typedef {{ file: string, mode: ChangeMode }} CommitChange
+ * @typedef {"A" | "C" | "D" | "M" | "R" | "T" | "U" | "X" | "B"} ChangeMode
+ */
+function getCommitInfo(line) {
+	const [, hash, message] = /^([a-f\d]+)\s+(.*)$/i.exec(line);
+
+	return git.raw(['diff-tree', '--no-commit-id', '--name-status', '-r', hash]).then(res => {
+		/* The output looks like this:
+		 *
+		 * M       components.js
+		 * M       components.json
+		 *
+		 * or nothing for e.g. reverts.
+		 */
+
+		const changes = !res ? [] : res.trim().split(/\n/g).map(line => {
+			const [, mode, file] = /(\w)\s+(.+)/.exec(line);
+			return { mode: /** @type {ChangeMode} */ (mode), file };
+		});
+		return { hash, message, changes };
+	});
+}
+
+/**
+ * Parses the output of `git log` with the given revision range.
+ *
+ * @param {string | Promise<string>} range The revision range in which the log will be parsed.
+ * @returns {Promise<CommitInfo[]>}
+ */
+function getLog(range) {
+	return Promise.resolve(range).then(range => {
+		return git.raw(['log', range, '--oneline']);
+	}).then(res => {
+		/* The output looks like this:
+		 *
+		 * bfbe4464 Invoke `callback` after `after-highlight` hook (#1588)
+		 * b41fb8f1 Fixes regex for JS examples (#1591)
+		 */
+		const commits = !res ? [] : res.trim().split(/\n/g);
+
+		return Promise.all(commits.map(getCommitInfo));
+	});
+}
+
+const revisionRanges = {
+	nextRelease: git.raw(['describe', '--abbrev=0', '--tags']).then(res => `${res.trim()}..HEAD`),
+	v1_15_0__HEAD: 'v1.15.0..HEAD',
+	v1_14_0__v1_15_0: 'v1.14.0..v1.15.0',
+};
+const strCompare = (a, b) => a.localeCompare(b, 'en');
+
+function changes() {
+
+	const { languages, plugins } = require('../components');
+
+	return getLog(revisionRanges.nextRelease).then(infos => {
+
+		const entries = {
+			'TODO:': {},
+			'New components': {
+				['']: createSortedArray(strCompare)
+			},
+			'Updated components': {},
+			'Updated plugins': {},
+			'Updated themes': {},
+			'Other': {},
+		};
+		/**
+		 *
+		 * @param {string} category
+		 * @param {string | { message: string, hash: string }} info
+		 */
+		function addEntry(category, info) {
+			const path = category.split(/\s*>>\s*/g);
+			if (path[path.length - 1] !== '') {
+				path.push('');
+			}
+
+			let current = entries;
+			for (const key of path) {
+				if (key) {
+					current = current[key] = current[key] || {};
+				} else {
+					(current[key] = current[key] || []).push(info);
+				}
+			}
+		}
+
+
+		/** @param {CommitChange} change */
+		function notGenerated(change) {
+			return !change.file.endsWith('.min.js') && ['prism.js', 'components.js', 'package-lock.json'].indexOf(change.file) === -1;
+		}
+		/** @param {CommitChange} change */
+		function notPartlyGenerated(change) {
+			return change.file !== 'plugins/autoloader/prism-autoloader.js' &&
```

Comment:
Will this also remove these files from manual changes to them?

Reply:
Yes and no.
When looking at where I used this function, you'll see that I always checked whether `components.json` was changed. I just assume that all changes in Autoloader & Show language are caused by the generated part of these files. This assumption isn't necessarily true but most of the time it is.

Only `addedComponent` and `changedComponentOrCore` use `notPartlyGenerated`. So you would have to change Autoloader and add/modify a language definition for the two to incorrectly categorize your PR.

It is possible that the sorters (if you know a better name, I'll take it) put a commit in the wrong category but it will rarely happen. But that's also the worst thing that can happen. No commits will be lost.

And, as I said before:
_Careful review of the generated changelog is still necessary._